### ISSUE TITLE
Performance regression in 1.11.x

### ISSUE THREAD

**Issue by dhedlund** [appsignal/appsignal-elixir#524]
## Describe the bug

There are a couple bugs that, when combined, produce a performance regression where ETS lookups consume high CPU.

1. There are some scenarios where records in the `:"$appsignal_transaction_registry"` never get cleaned up and the table continues to grow unbounded.  I have identified one scenario where I can reproduce this issue reliably by raising an exception from within an instrumented function.  This appears to have existed since before 1.11.  We have seen the number of records in our ETS tables reach into the hundreds of thousands due to some fallback error handling that relies on catching exceptions, where that exception catching occurs outside of the decorated function.

2. Once the ETS table has sizable number of records, calls to `Appsignal.TransactionRegistry.pids_and_monitor_references/1` become really slow.  When the process finishes, an attempt is made to remove the transaction from the ETS table.  In appsignal 1.0.13, it would scrub the ETS table, setting its size back to 0 afterward; not sure if this was a bug or if t was actively checking what was still in use.  In 1.11.x, the ETS table is not scrubbed and stays the same size; as a result, all subsequent calls to remove their process' transaction from the ETS table require scanning the large ETS table.  The new changes (as part of PR #505?) are probably good overall, but cause a performance regression in production environments until the ETS table growth issue can be properly addressed.

Side note: we were also seeing 100% error rates being reported to appsignal metrics once ETS table lookup performance dropped below a certain threshold, even though requests were completing successfully.  We haven't determined if this is a separate issue, or just due to something timing out from the above.

## To Reproduce

I've created a minimal repo that can reproduce both the ETS table growth and the performance issue we were seeing.  I added the following to my /etc/hosts file to blacklist appsignal domains so I wouldn't be spamming the service since it has to be active to repro.
```
127.0.0.1 appsignal.com
127.0.0.1 push.appsignal.com
```

1. Checkout and compile the following repo:
https://github.com/dhedlund/appsignal_hammer

2. In an iex shell:
```
# Reliably seed the appsignal ETS table by exercising a bug, note ETS table size of 250k
iex(1)> AppsignalHammer.seed(250_000)
[
  id: #Reference<0.1253646793.3556114433.22947>,
  read_concurrency: false,
  write_concurrency: true,
  compressed: false,
  memory: 7751447,
  owner: #PID<0.209.0>,
  heir: :none,
  name: :"$appsignal_transaction_registry",
  size: 250000,
  node: :nonode@nohost,
  named_table: true,
  type: :set,
  keypos: 1,
  protection: :public
]

# Spawn 100 processes that do nothing more than call into an instrumented appsignal
# function. It takes approximately 3 seconds to complete on my machine.
iex(2)> AppsignalHammer.hammer(100)     
3.250463

# Spawning all processes in parallel, it still takes 3-4 seconds.
iex(3)> AppsignalHammer.hammer(100, 100)
3.932907

# Note the ETS table size is still 250k
iex(4)> AppsignalHammer.ets_info()
[
  id: #Reference<0.1253646793.3556114433.22947>,
  read_concurrency: false,
  write_concurrency: true,
  compressed: false,
  memory: 7751447,
  owner: #PID<0.209.0>,
  heir: :none,
  name: :"$appsignal_transaction_registry",
  size: 250000,
  node: :nonode@nohost,
  named_table: true,
  type: :set,
  keypos: 1,
  protection: :public
]
```

On appsignal 1.10.13, the equivalent completes much faster:
```
# ETS table size is 250k, jut like on the newer version
iex(1)> AppsignalHammer.seed(250_000)
[
  id: #Reference<0.771501626.4091150343.127111>,
  read_concurrency: false,
  write_concurrency: true,
  compressed: false,
  memory: 7751447,
  owner: #PID<0.210.0>,
  heir: :none,
  name: :"$appsignal_transaction_registry",
  size: 250000,
  node: :nonode@nohost,
  named_table: true,
  type: :set,
  keypos: 1,
  protection: :public
]

iex(2)> AppsignalHammer.hammer(100)
0.261964

iex(3)> AppsignalHammer.hammer(100, 100)
0.00635

# Note the ETS table size is now 0
iex(4)> AppsignalHammer.ets_info()   
[
  id: #Reference<0.771501626.4091150343.127111>,
  read_concurrency: false,
  write_concurrency: true,
  compressed: false,
  memory: 1329,
  owner: #PID<0.210.0>,
  heir: :none,
  name: :"$appsignal_transaction_registry",
  size: 0,
  node: :nonode@nohost,
  named_table: true,
  type: :set,
  keypos: 1,
  protection: :public
]
```

**Comment by jeffkreeftmeijer** [comment]
Hey @dhedlund,

Thanks for taking the time to look into this, and thanks for your thorough description of the problem you're seeing. We're looking at why this is happening, and we'll update this issue when we know more. 

In the meantime, we've been working on a patch which doesn't use the TransactionRegistry anymore, in preparation of changes we're planning for the future. Instead, this branch uses the process dictionary to keep the reference to any process' transaction in the process itself. If you have time, could you try that and let us know how that works for you? 

```
{:appsignal, github: "appsignal/appsignal-elixir", branch: "transaction-dictionary"}
```

Alternatively, for anyone running into this, we recommend downgrading to 1.10.13 for now.

Thanks!

**Comment by jeffkreeftmeijer** [comment]
Hey @dhedlund,

Thanks once more for your detailed description and minimal example. We've been able to reproduce the issue you're seeing, and I've written a patch that should solve the issue while still using the registry to store transactions. Could you try the [https://github.com/appsignal/appsignal-elixir/tree/registry-monitors](registry-monitors branch), to verify that solves the issue for you? If so, we'll go ahead and release an update to squash this bug.

```
{:appsignal, github: "appsignal/appsignal-elixir", branch: "registry-monitors"}
```

**Comment by tombruijn** [comment]
@jeffkreeftmeijer will this be fixed by #525

**Comment by jeffkreeftmeijer** [comment]
We've just released version 1.11.2, which includes a patch for this issue (#525). Please update to this version if you're currently running 1.11.x, and please let us know if the patch fixed the issue for you.

@dhedlund thanks again for digging in and opening this well-described issue. Please let us know if you run into anything else. :)

**Comment by dhedlund** [comment]
Thanks for getting to this!  We rolled back to 1.10.12 before filing the issue, so we've been in a stable place.  We'll try to test out the new version next week.
